Xkcd-Comics/Services/UserDataService.swift

```diff
@@ -0,0 +1,62 @@
+//
+//  UserDataService.swift
+//  Xkcd-Comics
+//
+//  Created by Jia-Jiuan Tsai on 2022-02-25.
+//
+
+import Foundation
+
+class UserDataService {
+    private let userDefaults: UserDefaults
+    private let jsonEncoder: JSONEncoder
+    private let jsonDecoder: JSONDecoder
+    
+    init(userDefaults: UserDefaults, jsonEncoder: JSONEncoder, jsonDecoder: JSONDecoder) {
+        self.userDefaults = userDefaults
+        self.jsonEncoder = jsonEncoder
+        self.jsonDecoder = jsonDecoder
+    }
+
+    func addComicToFavorites(comic: Comic) {
+        guard var favoriteComics = favoriteComics() else {
+            print("addComicToFavorites: the list is empty")
+            encodeComics(comics: [comic])
+            return
+        }
+        favoriteComics.append(comic)
+        encodeComics(comics: favoriteComics)
+    }
+    
+    func favoriteComics() -> [Comic]? {
```

Comment:
But what happens when the userDefaults is empty/ nil?

Reply:
You'd still have the same guards and fallbacks, but instead of `return nil`,  you'd have `return []` and the method signature should be `-> [Comic]` with no `?`. That way, the caller of this method doesn't have to first check whether the array exists, and then what it contains, but only what it contains. 

The difference between empty array, no array saved yet, or user defaults key not found, are all implementation details of this class. These distinctions are not interesting outside of this class, and exposing them counts as _leaking implementation details_. All that matters for the caller is the actual list of favorites and what it contains.